jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java

```diff
@@ -0,0 +1,343 @@
+/*
+ * (C) Copyright 2003-2018, by Christoph Grüne and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.cycle;
+
+import org.jgrapht.Graph;
+import org.jgrapht.alg.util.Pair;
+import org.jgrapht.graph.GraphWalk;
+import org.jgrapht.util.TypeUtil;
+
+import java.util.*;
+
+/**
+ * Implementation of an algorithm for the local augmentation problem for the cyclic exchange neighborhood,
+ * i.e. it finds subset-disjoint negative cycles in a graph, based on
+ * Ravindra K. Ahuja, James B. Orlin, Dushyant Sharma,
+ * A composite very large-scale neighborhood structure for the capacitated minimum spanning tree problem,
+ * Operations Research Letters, Volume 31, Issue 3, 2003, Pages 185-194, ISSN 0167-6377,
+ * https://doi.org/10.1016/S0167-6377(02)00236-5. (http://www.sciencedirect.com/science/article/pii/S0167637702002365)
+ *
```

Comment:
Add a precise definition of subset-disjoint here, so that a user can understand the algorithm input/output specification without having to read the paper.  Also specify *which* such cycle it finds; i.e. is it the shortest by edge length?  least negative?  arbitrary?  etc.